### PR TITLE
 feat(modal): implement component

### PR DESCRIPTION
**Pull request by Leotheluck**
> A Modal component is used to overlay with important content the main view of a site and block interactions with it.

### CODE REVIEW

File: packages/ods/src/components/modal/tests/rendering/ods-modal.spec.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    describe('color', () => {
+      it('should be reflected', async() => {
+        const colorValue = ODS_MODAL_COLOR.warning;
+
+        await setup(`<ods-modal color="${colorValue}"></ods-modal>`);
+
+        expect(root?.getAttribute('color')).toBe(colorValue);
+      });
+
+      it('should be set to "information" by default', async() => {
```
Comment (dpellier) [inline packages/ods/src/components/modal/tests/rendering/ods-modal.spec.ts:28]: you can just put `to default value`, otherwise when default value change, it is very likely that the title doesn't get updated alongside the test

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.scss [hunk 2]
```diff
@@ -0,0 +1,122 @@
+@use '../../../../../style/focus';
+@use '../../../../../style/modal';
+@use '../../../../../style/input';
+
+@keyframes modal-open {
```
Comment (dpellier) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.scss:5]: move both to mixin to be shareable

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.scss [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+:host(.ods-modal--open) {
+  display: flex;
+}
+
+.ods-modal {
+  &__dialog {
+    @include modal.ods-modal();
+
+    animation: modal-open 0.2s ease-in-out;
```
Comment (dpellier) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.scss:54]: defined twice, also in the mixin

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/ods/react/tests/_app/src/components/ods-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React from 'react-dom/client';
+import { OdsModal, OdsText } from 'ods-components-react';
+
+const Modal = () => {
+  return (
+    <OdsModal isOpen>
```
Comment (aesteves60) [inline packages/ods/react/tests/_app/src/components/ods-modal.tsx:6]: I think we should test it with a trigger and call the open methods

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  render(): FunctionalComponent {
+    return (
+      <Host class={{
+        'ods-modal': true,
+        'ods-modal--open': this.isOpen,
+      }}>
+        <dialog class='ods-modal__dialog'
+          ref={(el) => this.modalDialog = el as HTMLDialogElement}
```
Comment (aesteves60) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx:88]:
```suggestion
          ref={ (el) => this.modalDialog = el as HTMLDialogElement }
```

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            [ `ods-modal__dialog__header--${this.color}` ]: true,
+          }}>
+            {
+              this.isDismissible &&
+              <button
+                class={{
+                  'ods-modal__dialog__header__close': true,
+                  [ `ods-modal__dialog__header__close--${this.color}` ]: true,
+                }}
+                onClick={ this.close.bind(this) }>
```
Comment (aesteves60) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx:106]: use an arrow function `() => this.close()`

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+  render(): FunctionalComponent {
+    return (
+      <Host class={{
+        'ods-modal': true,
+        'ods-modal--open': this.isOpen,
+      }}>
+        <dialog class='ods-modal__dialog'
+          ref={(el) => this.modalDialog = el as HTMLDialogElement}
+          onClick={(event) => {
```
Comment (aesteves60) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx:89]: add space and make oneline call

File: packages/storybook/stories/components/modal/modal.stories.ts
Comment (aesteves60) [inline packages/storybook/stories/components/modal/modal.stories.ts:1]: When I go to the docs, then to the Demo and reopen the docs the modal don't work and I have an error in the console
`Uncaught SyntaxError: redeclaration of const buttonActions`

File: packages/storybook/stories/components/modal/documentation.mdx [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+<Canvas of={ ModalStories.CustomCSS } sourceState='shown'/>
+
+# Examples
+
+## Default
+
+<Canvas of={ ModalStories.Default } sourceState='shown' />
+
+## Dismissible
```
Comment (aesteves60) [inline packages/storybook/stories/components/modal/documentation.mdx:36]: Not Dismissible

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                onClick={ this.close.bind(this) }>
+                <ods-icon name={ ODS_ICON_NAME.cross }>
+                </ods-icon>
+              </button>
+            }
+          </div>
+
+          <div class='ods-modal__dialog__content'>
+            <slot></slot>
+            <div class='ods-modal__dialog__content__actions'>
```
Comment (aesteves60) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx:115]: Maybe add a horizontal scroll when there are many button

**Review by aesteves60 — CHANGES_REQUESTED** [review]

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx (hunk 9 above)
Comment (Leotheluck) [reply packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx:115]: I believe this should be discussed with Design teams first

File: packages/ods/react/tests/e2e/ods-modal.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    let boundingBox = await modal?.boundingBox();
+
+    expect(boundingBox?.height).toBe(undefined);
+    expect(boundingBox?.width).toBe(undefined);
+
+    const button = await page.$('ods-button:not(:disabled)');
+
+    await button?.click();
+    // Small delay to ensure page console event has been resolved
+    await new Promise((resolve) => setTimeout(resolve, 100));
```
Comment (dpellier) [inline packages/ods/react/tests/e2e/ods-modal.e2e.ts:27]: not needed, you don't listen to the console

File: packages/ods/react/tests/e2e/ods-modal.e2e.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  });
+
+  it('renders the modal on button click', async() => {
+    const modal = await page.$('ods-modal');
+    let boundingBox = await modal?.boundingBox();
+
+    expect(boundingBox?.height).toBe(undefined);
+    expect(boundingBox?.width).toBe(undefined);
+
+    const button = await page.$('ods-button:not(:disabled)');
```
Comment (dpellier) [inline packages/ods/react/tests/e2e/ods-modal.e2e.ts:23]: just `await page.$('ods-button')`?

File: packages/storybook/stories/components/modal/documentation.mdx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+# Examples
+
+## Default
+
+<Canvas of={ ModalStories.Default } sourceState='shown' />
+
+## Not Dismissible
+
+<Canvas of={ ModalStories.Dismissible } sourceState='shown' />
```
Comment (dpellier) [inline packages/storybook/stories/components/modal/documentation.mdx:38]: Rename story also

File: packages/storybook/stories/components/modal/modal.stories.ts [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+export default meta;
+
+export const Demo: StoryObj = {
+  render: (args) => html`
+<ods-modal class="modal-demo" is-open="${args.isOpen}" is-dismissible="${args.isDismissible}" color="${args.color}">
+  ${unsafeHTML(args.content)}
+  ${unsafeHTML(args.actions)}
+</ods-modal>
+<style>
+  .modal-demo .my-text::part(text) {
```
Comment (dpellier) [inline packages/storybook/stories/components/modal/modal.stories.ts:24]: weird to have custom style (that user can't interact with) on a control that is customisable

File: packages/storybook/stories/components/modal/modal.stories.ts [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+
+<style>
+  .modal-custom-css .my-text {
+    color: red;
+  }
+</style>
+  `,
+};
+
+export const Dismissible: StoryObj = {
```
Comment (dpellier) [inline packages/storybook/stories/components/modal/modal.stories.ts:156]: rename

**Review by dpellier — CHANGES_REQUESTED** [review]
issue on doc:
- go to Demo, close the modal
- go to Documentation, click on custom CSS trigger button
=> modal does not open

File: packages/storybook/stories/components/modal/modal.stories.ts [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+const meta: Meta = {
+  title: 'ODS Components/User Feedback/Modal',
+  component: 'ods-modal',
+};
+
+export default meta;
+
+export const Demo: StoryObj = {
+  render: (args) => html`
```
Comment (astagnol) [inline packages/storybook/stories/components/modal/modal.stories.ts:18]: No ods-button for the Demo?

**Review by astagnol — CHANGES_REQUESTED** [review]
Capture d’écran 2024-05-31 à 14 47 27
Actions content is not displayed as expected in a "thinner" Modal

File: packages/storybook/stories/components/modal/modal.stories.ts [hunk 16]
```diff
@@ -14,12 +15,25 @@ const meta: Meta = {
...
+  const storybookRoot = document.querySelector('#storybook-root');
+  if (storybookRoot) {
+    storybookRoot.innerHTML = '';
+  }
+};
+
+const withResetRoot = (storyFn: any) => {
+  resetRoot();
+  return storyFn();
+};
```
Comment (dpellier) [inline packages/storybook/stories/components/modal/modal.stories.ts:28]: move `resetRoot` and `withResetRoot` on a dedicated `hook.ts` file (near `control.ts`)

File: packages/storybook/stories/components/modal/modal.stories.ts [hunk 17]
```diff
@@ -91,6 +105,7 @@ export const Demo: StoryObj = {
   margin-top: 1rem;
 }`
   },
+  decorators: [withResetRoot],
```
Comment (dpellier) [inline packages/storybook/stories/components/modal/modal.stories.ts:120]: apply only once on meta object instead:
```
const meta: Meta = {
  title: 'ODS Components/User Feedback/Modal',
  component: 'ods-modal',
  decorators: [withResetRoot],
};
```

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by feoche — CHANGES_REQUESTED** [review]
## Documentation
- [ ] Remove icons from Buttons
- [ ] Use only primary Buttons as triggers

## Demo
- [ ] As an example, set `max-width: 100%` on both action Buttons (for responsive purpose)
- [ ] First action Button should be in a `ghost` variant
- [ ] Buttons label should be "Cancel" and "Confirm"
- [ ] Remove icons from Buttons
- [ ] On Demo page load, closing Button is already focused
- [ ] When modifying controls, Modal flickers to gain focus each time

## Style
- [ ] Set a `1rem` gap between action Buttons

**Review by feoche — CHANGES_REQUESTED** [review]
## Style
- [x] Backdrop does not appear on Chromium browsers (see https://stackoverflow.com/a/63322762)
- [ ] When the modal gets opened, the user focus should move to the first focusable element and stay trapped inside it.

**Review by feoche — APPROVED** [review]

File: packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    } else {
+      this.isOpen = false;
+    }
+
+    this.odsModalClose.emit();
+  }
+
+  @Method()
+  async open(): Promise<void> {
+    document.body.style.overflow = 'hidden';
```
Comment (dpellier) [inline packages/ods/src/components/modal/src/components/ods-modal/ods-modal.tsx:45]: there is still an issue when the web-component is rendered with `isOpen` as it does not go through this.

Instead of using JS, maybe be safer to use css, try something like:
```
body:has(dialog[open]) {
  overflow: hidden;
}
```

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/storybook/stories/components/modal/migration.from.17.x.mdx [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+You can use the new `isOpen` attribute to obtain the same behavior.
+
+`dismissible` <img src="https://img.shields.io/badge/updated-00FFFF" />
+
+Has been updated.
+
+You can use the new `isDismissible` attribute to obtain the same behavior.
+
+`color` <img src="https://img.shields.io/badge/updated-00FFFF" />
```
Comment (dpellier) [inline packages/storybook/stories/components/modal/migration.from.17.x.mdx:30]: please order attr alphabetically

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by aesteves60 — APPROVED** [review]

**Review by dpellier — CHANGES_REQUESTED** [review]
Got an event worse behaviour with overflow, the CSS selector does not seems to work (tested on FF mac)
Please add a test about this overflow before fixing it anyway

File: packages/ods/src/components/modal/tests/rendering/ods-modal.e2e.ts [hunk 20]
```diff
@@ -107,6 +107,21 @@ describe('ods-modal rendering', () => {
 
       expect(isVisible).toBe(true);
     });
+
+    it('should set body overflow to hidden if set to true', async() => {
```
Comment (dpellier) [inline packages/ods/src/components/modal/tests/rendering/ods-modal.e2e.ts:111]: test also that the modal is actually visibile

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/ods/src/components/modal/tests/rendering/ods-modal.e2e.ts [hunk 21]
```diff
@@ -122,6 +122,40 @@ describe('ods-modal rendering', () => {
 
       expect(hasOverflowHiden).toBe(true);
     });
+
+    it('should have not a bounding box if unset', async() => {
```
Comment (dpellier) [inline packages/ods/src/components/modal/tests/rendering/ods-modal.e2e.ts:126]: not have

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by dpellier — APPROVED** [review]

**Review by astagnol — APPROVED** [review]
